qiskit/circuit/barrier.py

```diff
@@ -34,3 +35,7 @@ def broadcast_arguments(self, qargs, cargs):
 
     def c_if(self, classical, val):
         raise QiskitError('Barriers are compiler directives and cannot be conditional.')
+
+    def validate_parameter(self, parameter):
+        """Barrier paramenter"""
+        raise CircuitError("Barrier has no parameters: {0} was provided".format(type(parameter)))
```

Comment:
This one confuses me a bit, many gates do not have parameters why is this check only added for the Barrier? 
E.g. this is the current behaviour:
```python
>>> from qiskit.circuit import Barrier
>>> from qiskit.circuit.library import XGate
>>> b = Barrier(2)
>>> x = XGate()
>>> x.params = [0, 1]
>>> b.params = [0, 1]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jul/Work/Qiskit/qiskit-terra/qiskit/circuit/instruction.py", line 142, in params
    self._params.append(self.validate_parameter(single_param))
  File "/Users/jul/Work/Qiskit/qiskit-terra/qiskit/circuit/barrier.py", line 41, in validate_parameter
    raise CircuitError("Barrier has no parameters: {0} was provided".format(type(parameter)))
qiskit.circuit.exceptions.CircuitError: "Barrier has no parameters: <class 'int'> was provided"
```
I think we should either add this check to all gates w/o parameters or don't add it at all